plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java

```diff
@@ -1292,6 +1302,85 @@ public void dropColumn(ConnectorSession session, ConnectorTableHandle tableHandl
     }
 
+    @Override
+    public void setColumnType(ConnectorSession session, ConnectorTableHandle tableHandle, ColumnHandle columnHandle, Type type)
+    {
+        HiveTableHandle table = (HiveTableHandle) tableHandle;
+        failIfAvroSchemaIsSet(table);
+        HiveColumnHandle column = (HiveColumnHandle) columnHandle;
+
+        if (table.getPartitionNames().isPresent() && table.getPartitionNames().get().contains(column.getName())) {
+            throw new TrinoException(NOT_SUPPORTED, "Changing partition column types is not supported");
+        }
+        if (!canChangeColumnType(column.getType(), type)) {
+            throw new TrinoException(NOT_SUPPORTED, "Cannot change type from %s to %s".formatted(column.getType(), type));
+        }
+        metastore.setColumnType(table.getSchemaName(), table.getTableName(), column.getName(), toHiveType(type));
+    }
+
+    private static boolean canChangeColumnType(Type sourceType, Type newType)
+    {
+        if (sourceType.equals(newType)) {
+            return true;
+        }
+        if (sourceType == TINYINT) {
+            return newType == SMALLINT || newType == INTEGER || newType == BIGINT;
+        }
+        if (sourceType == SMALLINT) {
+            return newType == INTEGER || newType == BIGINT;
+        }
+        if (sourceType == INTEGER) {
+            return newType == BIGINT;
+        }
+        if (sourceType == REAL) {
+            return newType == DOUBLE;
+        }
+        if (sourceType instanceof VarcharType || sourceType instanceof CharType) {
+            return newType instanceof VarcharType || newType instanceof CharType;
+        }
+        if (sourceType instanceof DecimalType sourceDecimal && newType instanceof DecimalType newDecimal) {
+            return sourceDecimal.getScale() == newDecimal.getScale()
+                    && sourceDecimal.getPrecision() <= newDecimal.getPrecision();
+        }
+        if (sourceType instanceof ArrayType sourceArrayType && newType instanceof ArrayType newArrayType) {
+            return canChangeColumnType(sourceArrayType.getElementType(), newArrayType.getElementType());
+        }
+        if (sourceType instanceof RowType sourceRowType && newType instanceof RowType newRowType) {
+            List<RowType.Field> fields = Streams.concat(sourceRowType.getFields().stream(), newRowType.getFields().stream())
+                    .distinct()
+                    .collect(toImmutableList());
+            for (RowType.Field field : fields) {
+                String fieldName = field.getName().orElseThrow();
+                if (fieldExists(sourceRowType, fieldName) && fieldExists(newRowType, fieldName)) {
+                    if (!canChangeColumnType(
+                            findFieldByName(sourceRowType.getFields(), fieldName).getType(),
+                            findFieldByName(newRowType.getFields(), fieldName).getType())) {
+                        return false;
+                    }
```

Comment:
Dropping row fields is dangerous with Hive connector, because they can be later reintroduced with a different type (without compatibility check), thus indirectly allowing any type change between row fields.

Intentional?
(maybe we have to do that)

Reply:
That's intentional because dropping and adding the same columns (≠fields) with different types has the same issue either. I can disable the behavior in this PR though.